### PR TITLE
OSOE-150: Fix Powershell code

### PR DESCRIPTION
**Pull request by sarahelsaig**
OSOE-150

### CODE REVIEW

File: .github/actions/publish-nuget/New-NugetPackages.ps1 [hunk 1]
```diff
@@ -0,0 +1,18 @@
+param([array] $Arguments)
```
Comment (Piedone) [inline .github/actions/publish-nuget/New-NugetPackages.ps1]: This isn't used.

File: .github/actions/build-dotnet/action.yml [hunk 2]
```diff
@@ -97,6 +105,15 @@ runs:
         key: ${{ steps.detect-cache-configuration.outputs.key }}
         restore-keys: ${{ steps.detect-cache-configuration.outputs.restore-keys }}
 
+    - name: PowerShell Static Code Analysis
```
Comment (Piedone) [inline .github/actions/build-dotnet/action.yml]: We need to run this for PRs opened in this repo too.

File: .github/actions/build-dotnet/action.yml [hunk 3]
```diff
@@ -46,18 +46,26 @@ inputs:
...
       can be omitted. Otherwise make sure to specify it to avoid an "MSB1008: Only one project can be specified." error
-      that is thrown when the `*.sln` wildcard results in multiple files. The value is given to PowerShell's 
+      that is thrown when the `*.sln` wildcard results in multiple files. The value is given to PowerShell's
       `Get-ChildItem` cmdlet as-is, so grepping is still possible but a name with spaces must be escaped separately.
+  powershell-analyzer-path:
+    required: false
+    default: "./tools/Lombiq.Analyzers.PowerShell/Invoke-Analyzer.ps1"
+    description: >
+      The path of the "Invoke-Analyzer.ps1" file to be used for linting PowerShell script files. If not found, nothing
+      happens. So if you have no use for it, feel free to ignore.
```
Comment (Piedone) [inline .github/actions/build-dotnet/action.yml]: This is not related to .NET build. It can be part of the OC or .NET build workflow because we commonly use PS scripts in those projects too, but not in the .NET build action.

Instead, we can have a PS analyzer action and use that from the workflows.

Note the necessity to also update the PowerShell-Analyzers Readme related to this.

**Review by Piedone — CHANGES_REQUESTED** [review]

File: .github/actions/publish-nuget/New-NugetPackages.ps1 (hunk 1 above)
Comment (sarahelsaig) [reply .github/actions/publish-nuget/New-NugetPackages.ps1]: It is used on Line 10:
`dotnet pack $project -p:NuspecFile="$nuspecFile" @Arguments`

File: .github/actions/build-dotnet/action.yml (hunk 2 above)
Comment (sarahelsaig) [reply .github/actions/build-dotnet/action.yml]: It is ran on OSOCE and this repo is a submodule of it, so it's ran that way. Or do you mean something else?

File: .github/actions/publish-nuget/New-NugetPackages.ps1 (hunk 1 above)
Comment (Piedone) [reply .github/actions/publish-nuget/New-NugetPackages.ps1]: I mean, this whole script.

File: .github/actions/build-dotnet/action.yml (hunk 2 above)
Comment (Piedone) [reply .github/actions/build-dotnet/action.yml]: It feels a bit strange not to have a standalone build for this repo and depend on using it as a submodule (because you can work on it without OSOCE, as opposed to the OC projects) but I see the point.

File: .github/actions/publish-nuget/New-NugetPackages.ps1 (hunk 1 above)
Comment (sarahelsaig) [reply .github/actions/publish-nuget/New-NugetPackages.ps1]: [See commit description.](https://github.com/Lombiq/GitHub-Actions/pull/23/commits/6b19f3411fddcd3a05c25de03c9c59fe1f287beb) It's copied here to be analyzed and fixed, will be used over there.

Comment (Piedone) [reply .github/actions/publish-nuget/New-NugetPackages.ps1]: Please remove it once you're happy with it so it's only added in OSOE-110.

File: .github/actions/build-dotnet/action.yml (hunk 3 above)
Comment (sarahelsaig) [reply .github/actions/build-dotnet/action.yml]: My thought process was that we do static code analysis before build.

Comment (Piedone) [reply .github/actions/build-dotnet/action.yml]: We can do that, but PS analysis is unrelated to the .NET build, thus it shouldn't be in this action.

Comment (sarahelsaig) [reply .github/actions/build-dotnet/action.yml]: It's removed now due to using MSBuild only.

File: .github/workflows/test-analysis-failure.yml [hunk 4]
```diff
@@ -0,0 +1,99 @@
+name: Test Analysis Failure
```
Comment (Piedone) [inline .github/workflows/test-analysis-failure.yml:1]: Cool thing! Perhaps we should use this as a smoke test of .NET analyzers still working too under OSOE-110?

File: .github/workflows/test-analysis-failure.yml [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        default: 360
+        description: Configuration for the timeout-minutes parameter of the workflow. The 360 is GitHub's default.
+      build-expected-code-analysis-errors:
+        required: false
+        type: string
+        default: ""
+        description: >
+          If not empty, it indicates that the static code analysis should fail. Then each line should contain one error
+          code. The static code analysis should yield these and only these errors. This is ignored if 
+          `enable-code-analysis` is not "true". You may add a colon and then an explanation of the expected error.
```
Comment (Piedone) [inline .github/workflows/test-analysis-failure.yml]: Move this to the other build-* settings.

File: .github/actions/build-dotnet/Build-DotNetSolution.ps1 [hunk 6]
```diff
@@ -43,5 +50,51 @@ if (Test-Path src/Utilities/Lombiq.Gulp.Extensions/Lombiq.Gulp.Extensions.csproj
...
+Write-Output "Building solution with ``dotnet build $Solution $($buildSwitches -join " ")``."
+
+$errorLines = New-Object "System.Collections.Generic.List[string]"
+$errorCodes = New-Object "System.Collections.Generic.List[string]"
+
+$errorFormat = '^(.*)\((\d+),(\d+)\): error (.*)'
+dotnet build $Solution @buildSwitches 2>&1 | % {
+    if ($_ -notmatch $errorFormat) { return $_ }
+
+    ($null, $file, $line, $column, $message) = [regex]::Match($_, $errorFormat).Groups.Value
```
Comment (Piedone) [inline .github/actions/build-dotnet/Build-DotNetSolution.ps1]: Can be compiled.

File: .github/actions/build-dotnet/Build-DotNetSolution.ps1 [hunk 7]
```diff
@@ -43,5 +50,51 @@ if (Test-Path src/Utilities/Lombiq.Gulp.Extensions/Lombiq.Gulp.Extensions.csproj
...
+    $length = [System.Math]::Max($errorCodes.Count, $expectedErrorCodes.Count)
+    foreach ($index in 0..($length - 1))
+    {
+        $actual = $errorCodes[$index]
+        $expected = $expectedErrorCodes[$index]
+
+        if ($actual -eq $expected)
+        {
+            $report.AppendLine("#$index OK ($actual)") | Out-Null
+        } else
```
Comment (Piedone) [inline .github/actions/build-dotnet/Build-DotNetSolution.ps1]: Eh.

File: .github/actions/build-dotnet/Build-DotNetSolution.ps1 [hunk 8]
```diff
@@ -43,5 +50,51 @@ if (Test-Path src/Utilities/Lombiq.Gulp.Extensions/Lombiq.Gulp.Extensions.csproj
...
+    if ($noErrors) { Write-Output "::error file=$file,line=$line,col=$column::$message" }
+}
+
+if ($expectedErrorCodes)
+{
+    $errorCodes = $errorCodes | Sort-Object
+    $fail = 0
+    $report = New-Object "System.Text.StringBuilder" "`n"
+
+    $length = [System.Math]::Max($errorCodes.Count, $expectedErrorCodes.Count)
```
Comment (Piedone) [inline .github/actions/build-dotnet/Build-DotNetSolution.ps1:75]: This can be used to initialize the `StringBuilder` too.

**Review by Piedone — CHANGES_REQUESTED** [review]

File: .github/workflows/test-analysis-failure.yml (hunk 4 above)
Comment (sarahelsaig) [reply .github/workflows/test-analysis-failure.yml:1]: Yeah, I can do that once this is merged.

File: .github/actions/build-dotnet/Build-DotNetSolution.ps1 (hunk 6 above)
Comment (sarahelsaig) [reply .github/actions/build-dotnet/Build-DotNetSolution.ps1]: Sorry, what?

File: .github/actions/build-dotnet/Build-DotNetSolution.ps1 (hunk 7 above)
Comment (sarahelsaig) [reply .github/actions/build-dotnet/Build-DotNetSolution.ps1]: You have to do this bs in CLI, didn't think to reformat it afterwards.

File: .github/actions/build-dotnet/Build-DotNetSolution.ps1 (hunk 8 above)
Comment (sarahelsaig) [reply .github/actions/build-dotnet/Build-DotNetSolution.ps1:75]: I doubt it. According to [the docs](https://docs.microsoft.com/en-us/dotnet/api/system.text.stringbuilder.-ctor?view=net-6.0#system-text-stringbuilder-ctor(system-int32)), the `capacity` argument of `StringBuilder`'s constructor refers to the number of characters, while `$length` refers to the number of lines.

Comment (Piedone) [reply .github/actions/build-dotnet/Build-DotNetSolution.ps1:75]: Right.

File: .github/actions/build-dotnet/Build-DotNetSolution.ps1 (hunk 6 above)
Comment (Piedone) [reply .github/actions/build-dotnet/Build-DotNetSolution.ps1]: Yes!

Comment (Piedone) [reply .github/actions/build-dotnet/Build-DotNetSolution.ps1]: Talking about this: https://docs.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.regexoptions?view=net-6.0#:~:text=Fields-,Compiled,-8

Since this regex is executed for all lines of the output (i.e. a lot of times) but the expression doesn't change, this makes sense to use for perf.

Comment (sarahelsaig) [reply .github/actions/build-dotnet/Build-DotNetSolution.ps1]: image

**Review by Piedone — APPROVED** [review]
